### PR TITLE
Swift 2

### PR DESCRIPTION
**Pull request by nuclearace**
This will be merged immediately when Swift 2 is officially released.

### CODE REVIEW

File: SocketIOClientSwift/SocketEngine.swift [hunk 1]
```diff
@@ -246,63 +246,63 @@ public final class SocketEngine: NSObject, WebSocketDelegate, SocketLogClient {
...
                 this.waitingForPoll = false
-                
+
                 if this.fastUpgrade {
                     this.doFastUpgrade()
                 } else if !this.closed && this.polling {
                     this.doPoll()
                 }
-            }}.resume()
+            }}?.resume()
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketEngine.swift]: You probably don't want to ignore tasks that failed to be created, failing silently. Maybe change this to `!` instead?

File: SocketIOClientSwift/SocketEngine.swift [hunk 2]
```diff
@@ -310,80 +310,80 @@ public final class SocketEngine: NSObject, WebSocketDelegate, SocketLogClient {
...
-                
+
                 dispatch_async(this.emitQueue) {[weak this] in
                     if !(this?.fastUpgrade ?? true) {
                         this?.flushWaitingForPost()
                         this?.doPoll()
                     }
                 }
-            }}.resume()
+            }}?.resume()
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketEngine.swift]: Ditto

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,7 @@
 .DS_Store
 .AppleDouble
 .LSOverride
+*.xcodeproj
```
Comment (NachoSoto) [inline .gitignore]: Why this change? Looks like the project file is in the repo, which is extremely important (as explained in #117).

File: SocketIOClientSwift/SocketPacket.swift [hunk 4]
```diff
@@ -24,38 +24,17 @@
...
-        better = better["~~3"] ~= (placeholders != nil ? String(placeholders!) : "nil")
-        
-        return better
-    }
-    var id:Int?
-    var justAck = false
-    var nsp = ""
-    var placeholders:Int?
-    var type:PacketType?
+struct SocketPacket {
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketPacket.swift]: +1

**Comment by NachoSoto** [comment]
I took a quick look at the changes and everything looks good!

File: .gitignore (hunk 3 above)
Comment (nuclearace) [inline .gitignore]: If we add any new projects we'll have to manually add them.

**Comment by nuclearace** [comment]
@lightsprint09 any idea why the build is failing? I haven't made any radical code changes.

**Comment by nuclearace** [comment]
`Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'XCTestConfiguration isn't available'`

**Comment by nuclearace** [comment]
Also, at some point we should probably add tests for OS X

File: SocketIOClientSwift/SocketEngine.swift [hunk 5]
```diff
@@ -625,59 +592,59 @@ public final class SocketEngine: NSObject, WebSocketDelegate, SocketLogClient {
...
+
             postWait.append(strMsg)
-            
-            if datas != nil {
-                for data in datas! {
-                    let (nilData, b64Data) = createBinaryDataForSend(data)
-                    
+
+            if let datas = datas {
+                for data in datas {
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketEngine.swift]: You can simplify this by doing `for data in datas ?? []`. That way you reduce the indentation level.

File: SocketIOClientSwift/SocketIOClient.swift [hunk 6]
```diff
@@ -24,60 +24,42 @@
...
-    private var _connecting = false
-    private var currentReconnectAttempt = 0
-    private var handlers = ContiguousArray<SocketEventHandler>()
-    private var connectParams:[String: AnyObject]?
-    private var _secure = false
-    private var _reconnecting = false
-    private var reconnectTimer:NSTimer?
+public final class SocketIOClient: NSObject, SocketEngineClient {
+    public let emitQueue = dispatch_queue_create("emitQueue", DISPATCH_QUEUE_SERIAL)
+    public let handleQueue: dispatch_queue_t!
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: These 2 queues seem like an implementation detail, should they be `private`?

File: SocketIOClientSwift/SocketIOClient.swift [hunk 7]
```diff
@@ -24,60 +24,42 @@
...
     
+    private let logType = "SocketIOClient"
+    
+    private var anyHandler: ((SocketAnyEvent) -> Void)?
+    private var currentReconnectAttempt = 0
+    private var handlers = ContiguousArray<SocketEventHandler>()
+    private var connectParams: [String: AnyObject]?
+    private var reconnectTimer: NSTimer?
+    
+    let reconnectAttempts: Int!
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: This can be `private`?

File: SocketIOClientSwift/SocketIOClient.swift [hunk 8]
```diff
@@ -24,60 +24,42 @@
...
+    var ackHandlers = SocketAckManager()
+    var currentAck = -1
+    var waitingData = [SocketPacket]()
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: All these can be `private` too, right?

File: SocketIOClientSwift/SocketIOClient.swift [hunk 9]
```diff
@@ -86,24 +68,19 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
             self.connectParams = connectParams
         }
         
+        if let logger = opts?["logger"] as? SocketLogger {
+            Logger = logger
+        }
+        
         if let log = opts?["log"] as? Bool {
-            self.log = log
+            Logger.log = log
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: Seems like all files make use this singleton, but it gets set when creating a `SocketIOClient`? That's probably going to lead to some confusing behavior. I'd recommend not making it a singleton, or at least not making it mutable.

File: SocketIOClientSwift/SocketIOClient.swift [hunk 10]
```diff
@@ -155,13 +128,11 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
-    public func close(#fast:Bool) {
-        SocketLogger.log("Closing socket", client: self)
+    public func close(fast fast: Bool) {
+        Logger.log("Closing socket", type: logType)
         
         reconnects = false
-        _connecting = false
-        _connected = false
-        _reconnecting = false
+        status = SocketIOClientStatus.Closed
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: You can just do `.Closed` here

File: SocketIOClientSwift/SocketIOClient.swift [hunk 11]
```diff
@@ -175,42 +146,43 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
-        dispatch_after(time, dispatch_get_main_queue()) {[weak self] in
-            if let this = self where !this.connected {
-                this._closed = true
-                this._connecting = false
-                this.engine?.close(fast: true)
-                
-                handler?()
+    public func connect(timeoutAfter timeoutAfter:Int,
+        withTimeoutHandler handler:(() -> Void)?) {
+            guard status != SocketIOClientStatus.Connected else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: You can just do `.Connected`

File: SocketIOClientSwift/SocketIOClient.swift [hunk 12]
```diff
@@ -175,42 +146,43 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
+                return
+            }
+            if status == SocketIOClientStatus.Closed {
+                Logger.log("Warning! This socket was previously closed. This might be dangerous!",
+                    type: logType)
+            }
+            
+            status = SocketIOClientStatus.Connecting
+            addEngine()
+            engine?.open(connectParams)
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: I think this would be more correct as `engine!.open(connectParams)`. You probably don't want to ignore if `addEngine` didn't do the right thing.

File: SocketIOClientSwift/SocketIOClient.swift [hunk 13]
```diff
@@ -175,42 +146,43 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
-        let time = dispatch_time(DISPATCH_TIME_NOW, Int64(timeoutAfter) * Int64(NSEC_PER_SEC))
-        
-        dispatch_after(time, dispatch_get_main_queue()) {[weak self] in
-            if let this = self where !this.connected {
-                this._closed = true
-                this._connecting = false
-                this.engine?.close(fast: true)
-                
-                handler?()
+    public func connect(timeoutAfter timeoutAfter:Int,
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: Maybe you want to add `assert(timeoutAfter >= 0)` to make sure users don't pass negative values

File: SocketIOClientSwift/SocketIOClient.swift [hunk 14]
```diff
@@ -175,42 +146,43 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
     
-    private func createOnAck(event:String, items:[AnyObject]) -> OnAckCallback {
+    private func createOnAck(items: [AnyObject]) -> OnAckCallback {
         return {[weak self, ack = ++currentAck] timeout, callback in
             if let this = self {
                 this.ackHandlers.addAck(ack, callback: callback)
                 
                 dispatch_async(this.emitQueue) {[weak this] in
-                    this?._emit(event, items, ack: ack)
+                    this?._emit(items, ack: ack)
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: You don't need to make `this` `weak` here

File: SocketIOClientSwift/SocketIOClient.swift [hunk 15]
```diff
@@ -268,33 +234,27 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
     Same as emit, but meant for Objective-C
     */
-    public func emit(event:String, withItems items:[AnyObject]) {
-        if !connected {
+    public func emit(event: String, withItems items: [AnyObject]) {
+        guard status == SocketIOClientStatus.Connected else {
             return
         }
         
         dispatch_async(emitQueue) {[weak self] in
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/SocketIOClient.swift [hunk 16]
```diff
@@ -342,18 +291,15 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
-    func emitAck(ack:Int, withData args:[AnyObject]) {
+    func emitAck(ack: Int, withItems items: [AnyObject]) {
         dispatch_async(emitQueue) {[weak self] in
-            if let this = self where this.connected {
-                let packet = SocketPacket(type: nil, data: args, nsp: this.nsp, id: ack)
-                let str:String
-                
-                SocketParser.parseForEmit(packet)
-                str = packet.createAck()
+            if let this = self where this.status == SocketIOClientStatus.Connected {
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/SocketIOClient.swift [hunk 17]
```diff
@@ -387,15 +332,14 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
-    public func handleEvent(event:String, data:[AnyObject]?, isInternalMessage:Bool = false,
-        wantsAck ack:Int? = nil) {
-            // println("Should do event: \(event) with data: \(data)")
-            if !connected && !isInternalMessage {
+    public func handleEvent(event: String, data: [AnyObject]?, isInternalMessage: Bool,
+        wantsAck ack: Int? = nil) {
+            guard status == SocketIOClientStatus.Connected || isInternalMessage else {
                 return
             }
+            // println("Should do event: \(event) with data: \(data)")
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: Remove?

File: SocketIOClientSwift/SocketIOClient.swift [hunk 18]
```diff
@@ -432,27 +374,45 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
+    /**
     Removes handler(s)
     */
-    public func off(event:String) {
-        SocketLogger.log("Removing handler for event: %@", client: self, args: event)
+    public func off(event: String) {
+        Logger.log("Removing handler for event: %@", type: logType, args: event)
         
-        handlers = handlers.filter {$0.event == event ? false : true}
+        handlers = ContiguousArray(handlers.filter {!($0.event == event)})
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: You can just do `filter { $0.event != event }`

File: SocketIOClientSwift/SocketIOClient.swift [hunk 19]
```diff
@@ -480,53 +440,61 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
         connect()
     }
     
-    public func parseSocketMessage(msg:String) {
-        SocketParser.parseSocketMessage(msg, socket: self)
+    public func parseSocketMessage(msg: String) {
+        dispatch_async(handleQueue) {[weak self] in
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/SocketIOClient.swift [hunk 20]
```diff
@@ -480,53 +440,61 @@ public final class SocketIOClient: NSObject, SocketEngineClient, SocketLogClient
...
+            if let this = self {
+                SocketParser.parseSocketMessage(msg, socket: this)
+            }
+        }
     }
     
-    public func parseBinaryData(data:NSData) {
-        SocketParser.parseBinaryData(data, socket: self)
+    public func parseBinaryData(data: NSData) {
+        dispatch_async(handleQueue) {[weak self] in
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClient.swift]: Doesn't need to be `weak`

File: SocketIOClientSwift/SocketIOClientStatus.swift [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+import Foundation
+
+@objc public enum SocketIOClientStatus: Int, CustomStringConvertible {
+    public var description: String {
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketIOClientStatus.swift]: This isn't necessary in `Swift 2` anymore, it will create a description automatically.

File: SocketIOClientSwift/SocketPacket.swift [hunk 22]
```diff
@@ -94,128 +99,129 @@ final class SocketPacket: Printable {
...
+        for arg in data {
+            if arg is NSDictionary || arg is [AnyObject] {
+                do {
+                    let jsonSend = try NSJSONSerialization.dataWithJSONObject(arg,
+                        options: NSJSONWritingOptions(rawValue: 0))
+                    let jsonString = NSString(data: jsonSend, encoding: NSUTF8StringEncoding)
+                    
+                    message += jsonString! as String + ","
+                } catch {
+                    print("Error creating JSON object in SocketPacket.completeMessage")
```
Comment (NachoSoto) [inline SocketIOClientSwift/SocketPacket.swift]: Maybe use `Logger` here instead?

File: SocketIOClientSwift/SwiftRegex.swift [hunk 23]
```diff
@@ -18,34 +18,34 @@ var swiftRegexCache = [String: NSRegularExpression]()
...
                 self.regex = NSRegularExpression()
             }
         }
         super.init()
     }
-
+    
     class func failure(message: String) {
-        println("SwiftRegex: "+message)
+        print("SwiftRegex: "+message)
```
Comment (NachoSoto) [inline SocketIOClientSwift/SwiftRegex.swift]: `Logger`?

File: SocketIOClientSwift/WebSocket.swift [hunk 24]
```diff
@@ -133,13 +133,22 @@ public class WebSocket : NSObject, NSStreamDelegate {
         if isCreated {
             return
         }
-        dispatch_async(queue,{
-            self.didDisconnect = false
+
+        dispatch_async(queue, { [weak self] in
+            guard let weakSelf = self else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 25]
```diff
@@ -133,13 +133,22 @@ public class WebSocket : NSObject, NSStreamDelegate {
...
+            }
+
+            weakSelf.didDisconnect = false
         })
-        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0), {
-            self.isCreated = true
-            self.createHTTPRequest()
-            self.isCreated = false
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0), { [weak self] in
+            guard let weakSelf = self else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 26]
```diff
@@ -369,11 +378,16 @@ public class WebSocket : NSObject, NSStreamDelegate {
         }
         if totalSize > 0 {
             if validateResponse(buffer, bufferLen: totalSize) {
-                dispatch_async(queue,{
-                    if let connectBlock = self.onConnect {
+                dispatch_async(queue, { [weak self] in
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 27]
```diff
@@ -511,11 +525,15 @@ public class WebSocket : NSObject, NSStreamDelegate {
                 data = NSData(bytes: UnsafePointer<UInt8>((buffer+offset)), length: Int(len))
             }
             if receivedOpcode == OpCode.Pong.rawValue {
-                dispatch_async(queue,{
-                    if let pongBlock = self.onPong {
+                dispatch_async(queue, { [weak self] in
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 28]
```diff
@@ -599,24 +617,32 @@ public class WebSocket : NSObject, NSStreamDelegate {
...
+                let str: NSString? = NSString(data: response.buffer!, encoding: NSUTF8StringEncoding)
                 if str == nil {
                     writeError(CloseCode.Encoding.rawValue)
                     return false
                 }
-                dispatch_async(queue,{
-                    if let textBlock = self.onText {
-                        textBlock(str! as! String)
+                dispatch_async(queue, { [weak self] in
+                    guard let weakSelf = self else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 29]
```diff
@@ -599,24 +617,32 @@ public class WebSocket : NSObject, NSStreamDelegate {
...
+                        textBlock(str! as String)
+                    }
+                    weakSelf.delegate?.websocketDidReceiveMessage(weakSelf, text: str! as String)
                 })
             } else if response.code == .BinaryFrame {
                 let data = response.buffer! //local copy so it is perverse for writing
-                dispatch_async(queue,{
-                    if let dataBlock = self.onData {
+                dispatch_async(queue, { [weak self] in
+                    guard let weakSelf = self else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

File: SocketIOClientSwift/WebSocket.swift [hunk 30]
```diff
@@ -719,12 +745,16 @@ public class WebSocket : NSObject, NSStreamDelegate {
     ///used to preform the disconnect delegate
     private func doDisconnect(error: NSError?) {
         if !self.didDisconnect {
-            dispatch_async(queue,{
-                self.didDisconnect = true
-                if let disconnect = self.onDisconnect {
+            dispatch_async(queue, { [weak self] in
+                guard let weakSelf = self else {
```
Comment (NachoSoto) [inline SocketIOClientSwift/WebSocket.swift]: This doesn't need to be `weak`

**Comment by nuclearace** [comment]
It's best to default to `weak`. Especially for async.

**Comment by nuclearace** [comment]
@lightsprint09 I think the reason travis is failing is because of xctool. I think there's a bug in the 0.2.5 that they just upgraded to.

**Comment by lightsprint09** [comment]
I think so. I tried a lot of configurations this morning but non of them worked.

Also I found out that node is available in the OS X environment. I will try to fix travis and make the external server obsolete.

**Comment by bocodigital** [comment]
Is there a new cocoa pod for this yet?

**Comment by nuclearace** [comment]
@bocodigital I'll merge this into master and push a new pod once Xcode 7 is released fully.

**Comment by bocodigital** [comment]
ok thanks

**Comment by NachoSoto** [comment]
Cool I think this looks good!

I'm going to open another PR to add support for Apple TV soon, too :)

File: SocketIO-iOSTests/SocketParserTest.swift [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+        XCTAssertEqual(parser.readUntilStringOccurence("-"), "1")
+        XCTAssertEqual(parser.currentCharacter, "/")
+    }
+    
+    func validateParseResult(message: String) {
+        let validValues = SocketParserTest.packetTypes[message]!
+        let packet = SocketParser.parseString(message)
+        let type = message.substringWithRange(Range<String.Index>(start: message.startIndex, end: message.startIndex.advancedBy(1)))
+        if let packet = packet {
+            print(packet)
```
Comment (lightsprint09) [inline SocketIO-iOSTests/SocketParserTest.swift]: debug purpose?

Comment (nuclearace) [inline SocketIO-iOSTests/SocketParserTest.swift]: Forgot to remove it.

**Comment by nuclearace** [comment]
Going to merge this over since Xcode 7 is going to be coming out soon

**Comment by nuclearace** [comment]
@bocodigital There should be a new version on cocoapods
